src/Components/Components/src/Routing/TemplateSegment.cs

```diff
@@ -30,43 +29,55 @@ public TemplateSegment(string template, string segment, bool isParameter)
+                    // via the '?' token.
+                    var questionMarkIndex = Value.IndexOf('?');
+                    if (questionMarkIndex == Value.Length - 1)
+                    {
+                        IsOptional = true;
+                        Value = Value[0..^1];
+                    }
+                    // If the `?` optional marker shows up in the segment but not at the very end,
+                    // then throw an error.
+                    else if (questionMarkIndex >= 0)
+                    {
+                        throw new ArgumentException($"Malformed parameter '{segment}' in route '{template}'. '?' character can only appear at the end of parameter name.");
+                    }
+
+                    Constraints = Array.Empty<RouteConstraint>();
                 }
-                // If the `?` optional marker shows up in the segment but not at the very end,
-                // then throw an error.
-                else if (Value.IndexOf('?') >= 0 && Value.IndexOf('?') != Value.Length - 1)
+                else
                 {
-                    throw new ArgumentException($"Malformed parameter '{segment}' in route '{template}'. '?' character can only appear at the end of parameter name.");
-                }
+                    var tokens = Value.Split(':');
+                    if (tokens[0].Length == 0)
+                    {
+                        throw new ArgumentException($"Malformed parameter '{segment}' in route '{template}' has no name before the constraints list.");
+                    }
 
-                Constraints = Array.Empty<RouteConstraint>();
+                    Value = tokens[0];
+                    IsOptional = tokens[^1].EndsWith("?");
+                    if (IsOptional)
+                    {
+                        tokens[^1] = tokens[^1][0..^1];
+                    }
+
+                    Constraints = new RouteConstraint[tokens.Length - 1];
+                    for (var i = 1; i < tokens.Length; i++)
+                    {
+                        Constraints[i - 1] = RouteConstraint.Parse(template, segment, tokens[i]);
+                    }
+                }
             }
             else
             {
-                var tokens = Value.Split(':');
-                if (tokens[0].Length == 0)
-                {
-                    throw new ArgumentException($"Malformed parameter '{segment}' in route '{template}' has no name before the constraints list.");
-                }
-
-                // Set the IsOptional flag to true if any type constraints
-                // for this parameter are designated as optional.
-                IsOptional = tokens.Skip(1).Any(token => token.EndsWith("?"));
-
-                Value = tokens[0];
-                Constraints = tokens.Skip(1)
-                    .Select(token => RouteConstraint.Parse(template, segment, token))
-                    .ToArray();
+                Constraints = Array.Empty<RouteConstraint>();
```

Comment:
Maybe set the default value for `Constraints` before we enter the if-statement here.

Reply:
It's a bit weird, isn't it? since the other branch will have to set the constraints array to a different value.